crypto/libsodium-fork/src/libsodium/crypto_sign/ed25519/ref10/batch.c

```diff
@@ -0,0 +1,318 @@
+        heap_get_top2(heap, &max1, &max2, limbsize);
+
+        /* only one scalar remaining, we're done */
+        if (iszero256_modm_batch(heap->scalars[max2]))
+            break;
+
+        /* exhausted another limb? */
+        if (!heap->scalars[max1][limbsize])
+            limbsize -= 1;
+
+        /* can we extend to the 128 bit scalars? */
+        if (!extended && isatmost128bits256_modm_batch(heap->scalars[max1])) {
+            heap_extend(heap, count);
+            heap_get_top2(heap, &max1, &max2, limbsize);
+            extended = 1;
+        }
+
+        sub256_modm_batch(heap->scalars[max1], heap->scalars[max1], heap->scalars[max2], limbsize);
+        ge25519_cached cached_p;
+        ge25519_p1p1 p_as_p1p1;
+        ge25519_p3_to_cached(&cached_p,&heap->points[max1]);
+
+        ge25519_add(&p_as_p1p1, &heap->points[max2], &cached_p);
+        ge25519_p1p1_to_p3(&heap->points[max2], &p_as_p1p1);
+        heap_updated_root(heap, limbsize);
+    }
+
+    ge25519_multi_scalarmult_vartime_final(r, &heap->points[max1], heap->scalars[max1]);
+}
+
+/*
+* verifies ed25519 signatures in  batch. The algorithm is based on https://github.com/floodyberry/ed25519-donna 
+* implemention. we changed the algorithm according to https://eprint.iacr.org/2020/1244.pdf .
+* the batch size is between 2 and 64 signatures per batch. 
+* When the batch fails the function falls back to check signature one at a time.
+* the function returns 0 on success and fills and array of "valid" ints. 
+* 1 - for signature i passed verification
+* 0 - for signature i failed verification
+*/
+int crypto_sign_ed25519_open_batch(const unsigned char **m, const unsigned long long *mlen, const unsigned char **pk, const unsigned char **RS, size_t num, int *valid_p)
+{
+    batch_heap batch;
+    ge25519_p3  p;
+    ge25519_p3  p_mul;
+    sc25519 *r_scalars;
+    size_t i, batchsize;
+    unsigned char hram[64];
+    unsigned char batchsum[32];
+    int ret = 0;
+
+    for (i = 0; i < num; i++)
+        valid_p[i] = 1;
+
+    while (num >= 2) {
+        batchsize = (num > MAX_BATCH_SIZE) ? MAX_BATCH_SIZE : num;
+
+        /* validate the public key and signature */
+        for (i=0; i < batchsize; i++){
+            if (validate_ed25519_pk_and_sig(RS[i],pk[i]) != 0)
+                goto fallback;
+        }
+
+        /* generate r (scalars[batchsize+1]..scalars[2*batchsize] */
+        ed25519_randombytes_unsafe (batch.r, batchsize * 16);
+        
+
+        r_scalars = &batch.scalars[batchsize + 1];
+        for (i = 0; i < batchsize; i++)
+            expand256_modm(r_scalars[i], batch.r[i], 16);
+
+        /* compute scalars[0] = ((r1s1 + r2s2 + ...)) */
+        for (i = 0; i < batchsize; i++) {
+            expand256_modm(batch.scalars[i], RS[i] + 32, 32);
+            mul256_modm(batch.scalars[i], batch.scalars[i], r_scalars[i]);
+        }
+        for (i = 1; i < batchsize; i++)
+            add256_modm(batch.scalars[0], batch.scalars[0], batch.scalars[i]);
+
```

Comment:
I haven't spent enough time on this loop - but, would it be possible to merge this loops to take advantage of cache locality ?